README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mapfit

<!-- badges: start -->
<!-- badges: end -->

The goal of mapfit is to estimate parameters of phase-type distribution
(PH) and Markovian arrival process (MAP). PH/MAP fitting is required in
the analysis of non-Markovian models involving general distributions. By
replacing general distributions with estimated PH/MAP, we have
approximate the non-Markovian models with continuous-time Markov chains
(CTMCs). Our tool offers

-   PH/MAP fitting with grouped data
-   PH fitting with theoretical probability density functions.

These features help us to analyze non-Markovian models with phase
expansion.

## Installation

``` r
# Install devtools from CRAN
install.packages("mapfit")

# Or the development version from GitHub:
# install.packages("devtools")
devtools::install_github("okamumu/mapfit")
```

## PH Fitting

### Overview

PH distribution is defined as the time to absorption in a
time-homogeneous CTMC with an absorbing state. The p.d.f. and cumulative
distribution function (c.d.f.) are mathematically given as the
expressions using matrix exponential. Let
![\\boldsymbol{\\alpha}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Calpha%7D "\boldsymbol{\alpha}")
and
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}")
denote a probability (row) vector for determining an initial state and
an infinitesimal generator for transient states, respectively. Since the
c.d.f. is given by the probability that the current state of underlying
CTMC has already been absorbed, the c.d.f. of PH distribution is given
by

![
F(x) = 1 - \\boldsymbol{\\alpha} \\exp(\\boldsymbol{Q} x) \\boldsymbol{1},
](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%0AF%28x%29%20%3D%201%20-%20%5Cboldsymbol%7B%5Calpha%7D%20%5Cexp%28%5Cboldsymbol%7BQ%7D%20x%29%20%5Cboldsymbol%7B1%7D%2C%0A "
F(x) = 1 - \boldsymbol{\alpha} \exp(\boldsymbol{Q} x) \boldsymbol{1},
")

where
![\\boldsymbol{1}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B1%7D "\boldsymbol{1}")
is a column vector whose entries are 1. Also the p.d.f. can be obtained
by taking the first derivative of the c.d.f.;

![
f(x) = \\boldsymbol{\\alpha} \\exp(\\boldsymbol{Q} x) \\boldsymbol{\\xi},
](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%0Af%28x%29%20%3D%20%5Cboldsymbol%7B%5Calpha%7D%20%5Cexp%28%5Cboldsymbol%7BQ%7D%20x%29%20%5Cboldsymbol%7B%5Cxi%7D%2C%0A "
f(x) = \boldsymbol{\alpha} \exp(\boldsymbol{Q} x) \boldsymbol{\xi},
")

where
![\\boldsymbol{\\xi} = - \\boldsymbol{Q} \\boldsymbol{1}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Cxi%7D%20%3D%20-%20%5Cboldsymbol%7BQ%7D%20%5Cboldsymbol%7B1%7D "\boldsymbol{\xi} = - \boldsymbol{Q} \boldsymbol{1}").

The purpose of PH fitting is to determine PH parameters
![\\boldsymbol{\\alpha}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Calpha%7D "\boldsymbol{\alpha}")
and
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}")
so that the estimated PH distribution fits to observed data. There are
two different approaches; MM (moment match) method and MLE (maximum
likelihood estimation). The MM method is to find PH parameters whose
first few moments match to the moments from empirical data or
distribution functions. On the other hand, MLE is to find PH parameters
maximizing the likelihood (probability) of which the data is drawn from
the model as a sample.

### Data for PH fitting

The parameter estimation algorithm generally depends on data forms to be
used. mapfit deals with several kinds of data in PH fitting; point data,
weighted point data, grouped data, grouped data with missing values and
truncated data. The point data consists of independent and
identically-distributed (IID) random samples.

| Sample No. | Time |
|:----------:|:----:|
|     1      | 10.0 |
|     2      | 1.4  |
|     …      |  …   |
|    100     | 51.0 |

The above table shows an example of point data for a hundred IID
samples. The weighted point data is the point data in which all points
have their own weights. The weighted point data is used for numerical
integration of a density function in our tool. The grouped data consists
of break points and counts. For each of two successive break points, the
number of samples is counted as a bin. This is equivalent to the data
format to draw a histogram.

The grouped data with missing values allows us to use the grouped data
in which several counts are unknown (missing). In the tool, missing
counts are expressed by `NA`. Also, in the truncated data, several
samples are truncated at a point before their observations (right
censored data). The truncated data can be represented as the number of
samples in a specific interval from the point to infinity in the context
of grouped data.

| Time interval | Counts |
|:-------------:|:------:|
|   \[0, 10\]   |   1    |
|  \[10, 20\]   |  `NA`  |
|  \[20, 30\]   |   4    |
|  \[30, 40\]   |   10   |
|  \[40, 50\]   |  `NA`  |
|  \[50, 60\]   |   30   |
|  \[60, 70\]   |   10   |
|  \[70, 80\]   |   12   |
|  \[80, 90\]   |   4    |
|  \[90, 100\]  |   0    |
|  \[100, Inf)  |   5    |

The above table shows an example of the grouped data on break points 0,
10, 20, …, 100 where the data has missing values at the intervals
\[10,20\] and \[40,50\]. Furthermore, the last 5 samples are truncated
at 100.

| Time interval | Counts |
|:-------------:|:------:|
|   \[0, 10\]   |   1    |
|  \[10, 20\]   |  `NA`  |
|  \[20, 30\]   |   4    |
|  \[30, 40\]   |   10   |
|  \[40, 50\]   |  `NA`  |
|  \[50, 60\]   |   30   |
|  \[60, 70\]   |   10   |
|  \[70, 80\]   |   12   |
|  \[80, 90\]   |   4    |
|  \[90, 100\]  |   0    |
|  \[100, Inf)  |  `NA`  |

On the other hand, the aboe table shows an example of another grouped
data. In this case, several samples are truncated at 100 but we do not
know the exact number of truncated samples.

### Models and Methods

PH distributions are classified to sub-classes by the structure of
![\\boldsymbol{\\alpha}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Calpha%7D "\boldsymbol{\alpha}")
and
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}"),
and the parameter estimation algorithms depend on the class of PH
distribution. The tool deals with the following classes of PH
distribution:

-   general PH: The PH distribution in which there are no constraints on
    ![\\boldsymbol{\\alpha}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Calpha%7D "\boldsymbol{\alpha}")
    and
    ![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}").
    In the tool, this is referred to as \`ph’ class.
-   canonical form 1 (CF1): One of the minimal representations of
    acyclic PH distribution. The matrix
    ![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}")
    becomes a bidiagonal matrix whose entries are sorted in increasing
    order. In the tool, this is referred to as \`cf1’ class.
-   hyper-Erlang distribution: One of the representations of acyclic PH
    distribution. The distribution consists of a mixture of Erlang
    distributions. In the tool, this is referred to as \`herlang’ class.

The parameters of `ph` class are
![\\boldsymbol{\\alpha}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Calpha%7D "\boldsymbol{\alpha}"),
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}")
and
![\\boldsymbol{\\xi}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7B%5Cxi%7D "\boldsymbol{\xi}"),
which are defined as members, called *slots*, of S4 class in R. To
represent the matrix
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}"),
we use `Matrix` package which is an external package of R. The slots of
`cf1` class are inherited from the `ph` class. In addition to inherited
slots, `cf1` has a slot `rate` to store the absolute values of diagonal
elements of
![\\boldsymbol{Q}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cboldsymbol%7BQ%7D "\boldsymbol{Q}").
The `herlang` class has the slots for the mixed ratio, shape parameters
of Erlang components, rate parameters of Erlang components. Both of
`cf1` and `herlang` classes can be transformed to `ph` class by using
\`as’ method of R.

The R functions for PH parameter estimation are;

-   `phfit.point`: MLEs for general PH, CF1 and hyper-Erlang
    distribution from point and weighted point data. The estimation
    algorithms for general PH and CF1 are the EM algorithms proposed in
    \[1\]. The algorithm for hyper-Erlang distribution is the EM
    algorithm with shape parameter selection described in \[2,3\].
-   `phfit.group`: MLEs for general PH, CF1 and hyper-Erlang
    distribution from grouped and truncated data. The estimation
    algorithms for general PH and CF1 are the EM algorithms proposed in
    \[4\]. The algorithm for hyper-Erlang distribution is the EM
    algorithm with shape parameter selection, which is originally
    developed as an extension of \[2,3\].
-   `phfit.density`: MLEs for general PH, CF1 and hyper-Erlang
    distribution from a density function defined on the positive domain
    \[0, Inf). The function `phfit.density` calls `phfit.point` after
    making weighted point data. The weighted point data is generated by
    numerical quadrature. In the tool, the numerical quadrature is
    performed with a double exponential (DE) formula.
-   `phfit.3mom`: MM methods for acyclic PH distribution with the first
    three moments \[5,6\].

The functions `phfit.point`, `phfit.group` and `phfit.density` select an
appropriate algorithm depending on the class of a given PH distribution.
These functions return a list including the estimated model (`ph`, `cf1`
or `herlang`), the maximum log-likelihood (llf), Akaike information
criterion (aic) and other statistics of the estimation algorithm. Also,
the function `phfit.3mom` returns a `ph` class whose first three moments
match to the given three moments.

### Example

Here we introduce examples of the usage of PH fitting based on IID
samples from Weibull distribution. At first, we load the mapfit package
and generate IID samples from a Weibull distribution:

``` r
library(mapfit)

RNGkind(kind = "Mersenne-Twister")
set.seed(1234)
wsample <- rweibull(100, shape=2, scale=1)
```

`wsample` is set to a vector including a hundred IID samples generated
from Weibull distribution with scale parameter 1 and shape parameter 2.
`set.seed(1234)` means determing the seed of random numbers.

``` r
wsample
#>   [1] 1.47450394 0.68871906 0.70390766 0.68745901 0.38698715 0.66768398
#>   [7] 2.15798755 1.20774494 0.63744792 0.81550201 0.60487388 0.77911210
#>  [13] 1.12394405 0.28223484 1.10901777 0.42140013 1.11847354 1.14942511
#>  [19] 1.29542664 1.20832306 1.07241633 1.09317655 1.35593576 1.79415102
#>  [25] 1.23272029 0.45823831 0.80189104 0.29867185 0.42977942 1.75616647
#>  [31] 0.88603717 1.15209430 1.09019210 0.82379560 1.30718279 0.52428076
#>  [37] 1.26618210 1.16260990 0.08877248 0.46259605 0.76928163 0.66055079
#>  [43] 1.07949774 0.68927927 1.05326127 0.83015673 0.62445527 0.85066119
#>  [49] 1.18780419 0.51698994 1.61451826 1.08267929 0.57645512 0.82710123
#>  [55] 1.37015479 0.82783512 0.83982074 0.53486736 1.32097400 0.40547750
#>  [61] 0.38107465 1.78142911 1.07157783 2.07043978 1.19632106 0.58944015
#>  [67] 1.08505663 0.82231170 1.72143270 0.75610263 1.45189672 0.33667780
#>  [73] 2.05544853 0.49443698 1.55189411 0.80962054 0.97796650 1.63049390
#>  [79] 1.06650011 0.63460678 0.27649350 0.86658386 1.39556590 0.77994249
#>  [85] 1.27622488 0.32701528 0.97102624 1.08091533 1.35366256 0.33107018
#>  [91] 1.33917815 0.32386549 1.41750912 1.42403682 1.50035348 0.81868450
#>  [97] 1.09695465 1.90327589 1.08273771 0.54611792
```

Based on the point data, we can estimate PH parameters. Here we obtain
the estimated parameters for general PH with 5 states, CF1 with 5 states
and the hyper-Erlang with 5 states by the following commands,
respectively;

``` r
## phfit with GPH
phfit.point(ph=ph(5), x=wsample)
#> 
#> Maximum LLF: -59.436973
#> AIC: 176.873945
#> Iteration:  2000 / 2000
#> Computation time (user): 0.924000
#> Convergence: FALSE
#> Error (abs): 2.658070e-05 (tolerance Inf)
#> Error (rel): 4.472080e-07 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  8.354082e-11 0.9958273 0.0006050049 0.003567682 9.063406e-95 
#> Exit    :  0.1179994 3.90902e-93 6.464486e-06 0.05132675 4.831416 
#> Infinitesimal generator : 
#> 5 x 5 sparse Matrix of class "dgCMatrix"
#>                                                                           
#> [1,] -4.871469e+00  4.471403e-144   1.628003e-66  5.694646e-18  4.75347004
#> [2,]  6.303324e-02  -4.830687e+00   4.592402e+00  8.763450e-03  0.16648792
#> [3,]  1.236227e-04   9.657331e-18  -4.850744e+00  4.816559e+00  0.03405511
#> [4,]  4.455606e+00   3.099458e-66   8.686727e-21 -4.846059e+00  0.33912532
#> [5,]  4.366386e-15  3.315274e-225  3.603132e-124  9.071887e-51 -4.83141618

## phfit with CF1
phfit.point(ph=cf1(5), x=wsample)
#> Initializing CF1 ...
#> oxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> 
#> Maximum LLF: -59.416064
#> AIC: 136.832128
#> Iteration:  2000 / 2000
#> Computation time (user): 1.111000
#> Convergence: FALSE
#> Error (abs): 1.509990e-06 (tolerance Inf)
#> Error (rel): 2.541383e-08 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  0.8888552 0.003273583 0.08345855 0.02441269 1.479869e-101 
#> Rate    :  4.83095 4.83097 4.880507 4.88051 4.880513

## phfit with Hyper-Erlang
phfit.point(ph=herlang(5), x=wsample, ubound=3)
#> shape:  1 1 3  llf=-62.78
#> shape:  1 2 2  llf=-71.80
#> shape:  1 4  llf=-60.08
#> shape:  2 3  llf=-62.78
#> shape:  5  llf=-61.37
#> 
#> Maximum LLF: -60.083945
#> AIC: 128.167889
#> Iteration:  204 / 2000
#> Computation time (user): 0.034000
#> Convergence: TRUE
#> Error (abs): 8.852490e-07 (tolerance Inf)
#> Error (rel): 1.473354e-08 (tolerance 1.490116e-08)
#> 
#> Size : 2
#> Shape   :  1 4 
#> Initial :  0.01005939 0.9899406 
#> Rate    :  3.799481 4.058378
```

In the above example, the number of Erlang components is restructured to
3 or less by using `ubound` argument (see \[2\] in detail).

Also we present PH fitting with grouped data. In this example, we make
grouped data from the point data `wsample` by using the function `hist`
which is originally a function to draw a histogram.

``` r
h.res <- hist(wsample, breaks="fd", plot=FALSE)

h.res$breaks
#>  [1] 0.0 0.2 0.4 0.6 0.8 1.0 1.2 1.4 1.6 1.8 2.0 2.2

h.res$counts
#>  [1]  1  9 12 14 15 20 13  6  6  1  3
```

In the above, breaks are determined according to Freedman-Diaconis (FD)
rule. Then we can get estimated PH parameters from grouped data.

``` r
## phfit with GPH
phfit.group(ph=ph(5), counts=h.res$counts, breaks=h.res$breaks)
#> 
#> Maximum LLF: -22.812530
#> AIC: 103.625059
#> Iteration:  1647 / 2000
#> Computation time (user): 0.514000
#> Convergence: TRUE
#> Error (abs): 3.396318e-07 (tolerance Inf)
#> Error (rel): 1.488795e-08 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  8.191206e-08 1.395077e-64 0.9998728 0.0001250616 2.092792e-06 
#> Exit    :  8.185605e-06 4.88599 1.973519e-65 1.476447e-07 0.0001568522 
#> Infinitesimal generator : 
#> 5 x 5 sparse Matrix of class "dgCMatrix"
#>                                                                             
#> [1,]  -4.716760e+00  1.283349e-01   1.569076e-19  9.129134e-05  4.588326e+00
#> [2,]  8.528916e-123 -4.885990e+00  4.786686e-205  1.550664e-20  7.924260e-58
#> [3,]   4.349205e+00  3.927403e-03  -4.886069e+00  5.329321e-01  5.000110e-06
#> [4,]   1.528398e-55  4.846790e+00  8.559353e-115 -4.846790e+00  8.307741e-17
#> [5,]   6.936069e-20  2.830921e-05   1.273236e-61  4.685313e+00 -4.685498e+00

## phfit with CF1
phfit.group(ph=cf1(5), counts=h.res$counts, breaks=h.res$breaks)
#> Initializing CF1 ...
#> oxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> 
#> Maximum LLF: -22.811905
#> AIC: 63.623811
#> Iteration:  1304 / 2000
#> Computation time (user): 0.565000
#> Convergence: TRUE
#> Error (abs): 3.388332e-07 (tolerance Inf)
#> Error (rel): 1.485335e-08 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  0.8667713 2.82726e-05 0.1323529 0.0008475288 9.382448e-58 
#> Rate    :  4.684116 4.684117 4.886013 4.886015 4.886017

## phfit with Hyper-Erlang
phfit.group(ph=herlang(5), counts=h.res$counts, breaks=h.res$breaks)
#> shape:  1 1 1 1 1  llf=-61.14
#> shape:  1 1 1 2  llf=-35.55
#> shape:  1 1 3  llf=-26.45
#> shape:  1 2 2  llf=-35.55
#> shape:  1 4  llf=-23.59
#> shape:  2 3  llf=-26.45
#> shape:  5  llf=-23.97
#> 
#> Maximum LLF: -23.588309
#> AIC: 55.176617
#> Iteration:  204 / 2000
#> Computation time (user): 0.043000
#> Convergence: TRUE
#> Error (abs): 3.463910e-07 (tolerance Inf)
#> Error (rel): 1.468486e-08 (tolerance 1.490116e-08)
#> 
#> Size : 2
#> Shape   :  1 4 
#> Initial :  0.002267234 0.9977328 
#> Rate    :  3.186197 4.05926
```

Next we present the case where PH parameters are estimated from a
density function. The density function of Weibull distribution is given
by a function `dweibull`. Then we can also execute the following
commands;

``` r
## phfit with GPH
phfit.density(ph=ph(5), f=dweibull, shape=2, scale=1)
#> 
#> Maximum LLF: -11.251992
#> AIC: 80.503984
#> Iteration:  2000 / 2000
#> Computation time (user): 0.830000
#> Convergence: FALSE
#> Error (abs): 6.961001e-06 (tolerance Inf)
#> Error (rel): 6.186457e-07 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  4.577855e-49 0.01368209 0.9497458 0.03657207 8.584149e-09 
#> Exit    :  4.946403 0.0001532047 1.150315e-45 3.097348e-05 0.2731335 
#> Infinitesimal generator : 
#> 5 x 5 sparse Matrix of class "dgCMatrix"
#>                                                                         
#> [1,] -4.9464112  3.319514e-65  3.208540e-105  2.649815e-27  7.716337e-06
#> [2,]  0.2559562 -4.961494e+00   7.518925e-10  4.697923e+00  7.462019e-03
#> [3,]  0.5438509  4.116477e+00  -4.864664e+00  1.087195e-01  9.561702e-02
#> [4,]  0.1824197  2.980250e-13   2.459518e-32 -4.929520e+00  4.747069e+00
#> [5,]  4.7184058  1.030910e-42   9.896035e-77  2.082164e-14 -4.991539e+00

## phfit with CF1
phfit.density(ph=cf1(5), f=dweibull, shape=2, scale=1)
#> Initializing CF1 ...
#> oxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> 
#> Maximum LLF: -11.247614
#> AIC: 40.495228
#> Iteration:  2000 / 2000
#> Computation time (user): 1.031000
#> Convergence: FALSE
#> Error (abs): 2.792283e-07 (tolerance Inf)
#> Error (rel): 2.482556e-08 (tolerance 1.490116e-08)
#> 
#> Size : 5
#> Initial :  0.7536077 0.003014352 0.1516817 0.09169629 2.445755e-37 
#> Rate    :  4.844509 4.844531 5.062429 5.062435 5.06244

## phfit with Hyper-Erlang
phfit.density(ph=herlang(5), f=dweibull, shape=2, scale=1)
#> shape:  1 1 1 1 1  llf=-16.55
#> shape:  1 1 1 2  llf=-12.44
#> shape:  1 1 3  llf=-11.48
#> shape:  1 2 2  llf=-12.44
#> shape:  1 4  llf=-11.39
#> shape:  2 3  llf=-11.51
#> shape:  5  llf=-12.83
#> 
#> Maximum LLF: -11.391140
#> AIC: 30.782281
#> Iteration:  76 / 2000
#> Computation time (user): 0.043000
#> Convergence: TRUE
#> Error (abs): 1.461348e-07 (tolerance Inf)
#> Error (rel): 1.282881e-08 (tolerance 1.490116e-08)
#> 
#> Size : 2
#> Shape   :  1 4 
#> Initial :  0.07425929 0.9257407 
#> Rate    :  2.132549 4.349237
```

The last two arguments for each execution are parameters of `dweibull`
function. User-defined functions are also used as density functions in
similar manner.

Usually, the PH fitting with density is used for the PH expansion (PH
approximation) in which known general desitributions are replaced with
the PH distributions estimated from these density functions. Compared to
PH fitting with samples, PH fitting with density function tends to be
accurate, because density function has more information than samples.
Therefore, in the case of PH fitting with density function, we can treat
PH distributions with high orders without causing overfitting, i.e., it
is possible to perform PH fitting even if PH has 100 states;

``` r
## estimate PH parameters from the density function
(result.density <- phfit.density(ph=cf1(100), f=dweibull, shape=2, scale=1))
#> Initializing CF1 ...
#> ooooxx
#> xxxxox
#> xxooxx
#> xxooxx
#> xxxxxx
#> xxxxxx
#> 
#> Maximum LLF: -11.208672
#> AIC: 420.417344
#> Iteration:  18 / 2000
#> Computation time (user): 0.322000
#> Convergence: TRUE
#> Error (abs): 1.427159e-07 (tolerance Inf)
#> Error (rel): 1.273264e-08 (tolerance 1.490116e-08)
#> 
#> Size : 100
#> Initial :  2.599883e-06 6.238748e-06 1.39668e-05 2.932778e-05 5.806999e-05 0.0001089792 0.0001947346 0.0003326876 0.0005453356 0.0008609362 0.001313022 0.001940568 0.002752387 0.003779144 0.005034756 0.006522314 0.008232693 0.01014416 0.01222305 0.01442546 0.01669971 0.01898941 0.02123668 0.02338547 0.02538437 0.027189 0.02876368 0.03008234 0.03112875 0.03189611 0.03238624 0.03260834 0.03257765 0.03231402 0.03184055 0.03118232 0.03036529 0.02941539 0.02835779 0.02721633 0.02601311 0.02476824 0.02349969 0.02222323 0.02095244 0.01969885 0.01847198 0.01727957 0.01612771 0.01502101 0.01396282 0.01295535 0.01199987 0.01109684 0.01024606 0.009446795 0.008697862 0.00799775 0.007344687 0.006736716 0.006171752 0.005647633 0.005162157 0.004713123 0.004298349 0.003915701 0.003563102 0.003238551 0.002940127 0.002666 0.002414428 0.002183764 0.001972456 0.001779041 0.001602152 0.001440505 0.001292906 0.001158241 0.001035476 0.0009236501 0.0008218768 0.000729335 0.0006452666 0.0005689707 0.0004997979 0.0004371446 0.0003804479 0.0003291814 0.0002828541 0.0002410142 0.0002032566 0.000169236 0.000138679 0.0001113854 8.719624e-05 6.589667e-05 4.700055e-05 2.941743e-05 1.195204e-05 5.344346e-07 
#> Rate    :  14.78982 14.79032 14.79221 14.79361 14.79503 14.79602 14.79734 14.7997 14.80516 14.80938 14.81343 14.81371 14.94164 15.10413 15.30104 15.53234 15.79787 16.09746 16.43096 16.7983 17.19949 17.6347 18.10423 18.60851 19.14814 19.72383 20.33645 20.98694 21.6764 22.40598 23.17695 23.99067 24.84856 25.75216 26.70307 27.70298 28.75369 29.85708 31.01512 32.2299 33.5036 34.83852 36.23704 37.70168 39.23509 40.84001 42.51933 44.27606 46.11334 48.03448 50.0429 52.1422 54.33613 56.6286 59.0237 61.52569 64.13904 66.86838 69.71858 72.69468 75.80199 79.04601 82.43248 85.96743 89.6571 93.50804 97.52705 101.7213 106.0981 110.6653 115.4309 120.4033 125.5914 131.0043 136.6516 142.5432 148.6896 155.1015 161.7904 168.768 176.0467 183.6393 191.5592 199.8203 208.4373 217.4254 226.8004 236.5789 246.7781 257.416 268.5113 280.0837 292.1536 304.7422 317.8719 331.5657 345.8478 360.743 376.277 392.4783
```

The result provides a highly-accurate approximation for Weibull
distribution. However, from the viewpoint of computation time, it should
be noted that only `cf1` or `herlang` with lower/upper bounds of Erlang
components can be applied to PH fitting with high orders. In the above
example, although the number of states is 100, the execution time is in
a few seconds because of the refinement of EM algorithm \[1\].

If we use only point data to estimate PH paramters with high orders, the
overfitting is happen.

``` r
## estimate PH parameters from 100 samples (overfitting example)
(result.point <- phfit.point(ph=cf1(100), x=wsample))
#> Initializing CF1 ...
#> oooxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> xxxxxx
#> 
#> Maximum LLF: -50.504694
#> AIC: 499.009389
#> Iteration:  2000 / 2000
#> Computation time (user): 13.911000
#> Convergence: FALSE
#> Error (abs): 2.206856e-05 (tolerance Inf)
#> Error (rel): 4.369603e-07 (tolerance 1.490116e-08)
#> 
#> Size : 100

NEWS.md
# mapfit 0.9.9

- Added a `NEWS.md` file to track changes to the package.
- Change the license
- Use Roxygen2





src/gamma.c
/*
Gamma Functions
*/

#include <math.h>

static int FACTMAX = 20;
static double PI = 3.14159265358979324; // pi
static double LOG_2PI = 1.83787706640934548;
static double LOG_PI = 1.14472988584940017; // log(pi)

static int N = 8;
//static double B0 = 1;
//static double B1 = (-1.0 / 2.0);
static double B2 = ( 1.0 / 6.0);
static double B4 = (-1.0 / 30.0);
static double B6 = ( 1.0 / 42.0);
static double B8 = (-1.0 / 30.0);
static double B10 = ( 5.0 / 66.0);
static double B12 = (-691.0 / 2730.0);
static double B14 = ( 7.0 / 6.0);
static double B16 = (-3617.0 / 510.0);

static double nfact[] = {
1.0,                        // 0
1.0,                        // 1
2.0,                        // 2
6.0,                        // 3
24.0,                       // 4
120.0,                      // 5
720.0,                      // 6
5040.0,                     // 7
40320.0,                    // 8
362880.0,                   // 9
3628800.0,                  // 10
39916800.0,                 // 11
479001600.0,                // 12
6227020800.0,               // 13
87178291200.0,              // 14
1307674368000.0,            // 15
20922789888000.0,           // 16
355687428096000.0,          // 17
6402373705728000.0,         // 18
121645100408832000.0,       // 19
2432902008176640000.0       // 20
};

static double lognfact[] = {
  0.0,
  0.0,
  0.6931471805599453,
  1.791759469228055,
  3.1780538303479458,
  4.787491742782046,
  6.579251212010101,
  8.525161361065415,
  10.60460290274525,
  12.801827480081469,
  15.104412573075516,
  17.502307845873887,
  19.987214495661885,
  22.552163853123425,
  25.19122118273868,
  27.89927138384089,
  30.671860106080672,
  33.50507345013689,
  36.39544520803305,
  39.339884187199495,
  42.335616460753485
};

double mylgamma(double x) {
  double v, w;
  v = 1;
  while (x<N) { v *=x; x++; }
  w = 1 / (x * x);
  return ((((((((B16 / (16 * 15)) * w + (B14 / (14 * 13))) * w
    + (B12 / (12 * 11))) * w + (B10 / (10 * 9))) * w
  + (B8 / (8 * 7))) * w + (B6 / (6 * 5))) * w
  + (B4 / (4 * 3))) * w + (B2 / (2 * 1))) / x
  + 0.5 * LOG_2PI - log(v) - x + (x - 0.5) * log(x);
}

double mytgamma(double x) {
  if (x < 0) {
    return PI / (sin(PI * x) * exp(mylgamma(1-x)));
  }
  return exp(mylgamma(x));
}


double psi(double x) {
  double v, w;
  v = 0;
  while (x < N) { v += 1 / x; x++; }
  w = 1 / (x * x);
  v += ((((((((B16 / 16) * w + (B14 /14)) * w
    + (B12 / 12)) * w + (B10 / 10)) * w
  + (B8 / 8)) * w + (B6 / 6)) * w
  + (B4 / 4)) * w + (B2 / 2)) * w + 0.5 / x;
  return log(x) - v;
}

double polygamma(int n, double x) {
  int k;
  double t, u, v, w;
  u = 1;
  for(k=1-n; k<0; k++) u *= k;
    v = 0;
  while (x<N) { v +=1 / pow(x, n+1); x++; }
  w = x * x;
  t = (((((((B16
    * (n + 15.0) * (n + 14) / (16 * 15 * w) + B14)
  * (n + 13.0) * (n + 12) / (14 * 13 * w) + B12)
  * (n + 11.0) * (n + 10) / (12 * 11 * w) + B10)
  * (n + 9.0) * (n + 8) / (10 * 9 * w) + B8)
  * (n + 7.0) * (n + 6) / (8 * 7 * w) + B6)
  * (n + 5.0) * (n + 4) / (6 * 5 * w) + B4)
  * (n + 3.0) * (n + 2) / (4 * 3 * w) + B2)
  * (n + 1.0) * n / (2 * 1 * w)
  + 0.5 * n / x + 1;
  return u * (t / pow(x, n) + n * v);
}

double tfact(int s) {
  if (s <= FACTMAX) {
    return nfact[s];
  } else {
    return exp(mylgamma(1.0 + s));
  }
}

double lfact(int s) {
  if (s <= FACTMAX) {
    return lognfact[s];
  } else {
    return mylgamma(1.0 + s);
  }
}

